### PR TITLE
fix: listen shadowsocks with TLS if configuration is provided

### PR DESCRIPTION
**Pull request by WendelHime**

### CODE REVIEW

File: http_proxy.go [hunk 1]
```diff
@@ -826,9 +826,28 @@ func (p *Proxy) listenShadowsocks(addr string) (net.Listener, error) {
...
+	if tlsConfig != nil {
+		base, err = tls.Listen("tcp", addr, tlsConfig)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		base, err = net.Listen("tcp", addr)
+		if err != nil {
+			return nil, err
+		}
```
Comment (WendelHime) [inline http_proxy.go:850]: Do we want to keep the support if the TLS configuration is not provided? If we don't I'll remove this piece

File: http_proxy.go [hunk 2]
```diff
@@ -826,9 +826,28 @@ func (p *Proxy) listenShadowsocks(addr string) (net.Listener, error) {
...
+		if err != nil {
+			return nil, errors.New("unable to load cert: %v", err)
+		}
+
+		tlsConfig = &tls.Config{Certificates: []tls.Certificate{cert}}
+	}
+
+	var base net.Listener
+	if tlsConfig != nil {
+		base, err = tls.Listen("tcp", addr, tlsConfig)
```
Comment (hwh33) [inline http_proxy.go:842]: This is backwards.  The TLS layer needs to be on top of the Shadowsocks layer.  Using TLS as the transport underneath Shadowsocks, as done here, will result in TLS being visible on the wire, not Shadowsocks.  I think using `tls.NewListener` and wrapping the Shadowsocks listener will give us the desired result.

The ordering of the network stack gets confusing, especially when we do weird stuff like this and compose different application-layer protocols.

File: http_proxy.go (hunk 1 above)
Comment (hwh33) [reply http_proxy.go:850]: I like the way you've done it!  This will give us the ability to run plain Shadowsocks if we ever want to.

**Review by hwh33 — COMMENTED** [review]
Nice work @WendelHime!  We just need to switch the order of the two protocols, but otherwise this is looking great.

File: http_proxy.go (hunk 2 above)
Comment (WendelHime) [reply http_proxy.go:842]: Aaaaaah got it! Thanks for explaining!

**Review by hwh33 — APPROVED** [review]
This looks great to me @WendelHime!  As discussed yesterday, we will wait until we've tested this before merging.

**Comment by WendelHime** [comment]
Converting to draft just to make sure

**Pull request merged**